caffe2/operators/quantized/int8_sigmoid_op.h

```diff
@@ -1,219 +1,99 @@
-    }
-    output_data[c] = output_val;
-  }
-}
-
-} // namespace
 
 class Int8SigmoidOp final : public Operator<CPUContext> {
  public:
   Int8SigmoidOp(const OperatorDef& operator_def, Workspace* ws)
-      : Operator<CPUContext>(operator_def, ws) {}
+      : Operator<CPUContext>(operator_def, ws),
+        ws_(ws) {}
+
+  ~Int8SigmoidOp() {
+    if (this->qnnpackOperator_ != nullptr) {
+      qnnp_delete_operator(this->qnnpackOperator_);
+      this->qnnpackOperator_ = nullptr;
+    }
+  }
 
   bool RunOnDevice() override {
     const auto& X = Inputs()[0]->template Get<Int8TensorCPU>();
     auto* Y = Outputs()[0]->template GetMutable<Int8TensorCPU>();
-    const int32_t Y_offset =
+    const int32_t Y_zero_point =
         this->template GetSingleArgument<int>("Y_zero_point", 0);
-    const float Y_scale = this->template GetSingleArgument<float>("Y_scale", 1);
-    CHECK_EQ(Y_offset, 0);
-    CHECK_EQ(Y_scale, 1. / 256);
+    const float Y_scale =
+        this->template GetSingleArgument<float>("Y_scale", 1);
+    CHECK_EQ(Y_zero_point, 0);
+    CHECK_EQ(Y_scale, 1.0f / 256.0f);
+
+    /*
+     * Record quantization parameters for the input, because if the op is
+     * in-place, we may overwrite these parameters later, when we set
+     * quantization parameters for output tensor.
+     */
+    const uint8_t X_zero_point = X.zero_point;
+    const float X_scale = X.scale;
 
     Y->scale = Y_scale;
-    Y->zero_point = Y_offset;
+    Y->zero_point = Y_zero_point;
     Y->t.ResizeLike(X.t);
-    int32_t input_multiplier;
-    int input_left_shift;
-    int input_range_radius =
-        SigmoidPrepare(X.scale, &input_multiplier, &input_left_shift);
-    Int8Logistic(
-        X.t.data<uint8_t>(),
-        Y->t.mutable_data<uint8_t>(),
-        X.zero_point,
-        input_range_radius,
-        input_multiplier,
-        input_left_shift,
-        X.t.numel() / X.t.size(0));
+
+    initQNNPACK();
+
+    pthreadpool_t threadpool =
+        reinterpret_cast<pthreadpool_t>(ws_->GetThreadPool());
+
+    if (this->qnnpackOperator_ == nullptr) {
+      const qnnp_status createStatus = qnnp_create_sigmoid_nc_q8(
+        1 /* channels */,
+        X_zero_point, X_scale,
+        static_cast<uint8_t>(Y_zero_point), Y_scale,
+        0 /* output min */,
+        255 /* output max */,
+        &qnnpackOperator_);
+      CAFFE_ENFORCE(
+          createStatus == qnnp_status_success,
+          "failed to create QNNPACK Sigmoid operator");
+      CAFFE_ENFORCE(this->qnnpackOperator_ != nullptr);
+    }
+
+    const qnnp_status setupStatus = qnnp_setup_sigmoid_nc_q8(
+        this->qnnpackOperator_,
+        X.t.numel() /* batch size */,
+        X.t.template data<uint8_t>(),
+        1 /* X stride */,
+        Y->t.template mutable_data<uint8_t>(),
+        1 /* Y stride */);
+    CAFFE_ENFORCE(
+        setupStatus == qnnp_status_success,
+        "failed to setup QNNPACK Sigmoid operator");
+
+#ifdef FBCODE_CAFFE2
```

Comment:
Can you explain this one quickly? Why don't we use the threadpool in fbcode?

Reply:
Because both `pthreadpool` and Caffe2 thread pool implement the same functions, and fbcode does not allow dependency on interface without dependency on implementation. QNNPACK depends on pthreadpool interface, therefore in fbcode it depends on pthreadpool implementation, and therefore we can't just pass Caffe2 thread pool pointer here.

Reply:
Thanks for the clarification. Given that, do you want to move

```
    pthreadpool_t threadpool =
        reinterpret_cast<pthreadpool_t>(ws_->GetThreadPool());
```
down here as well?

Reply:
yes, good point